libirods_rule_engine_plugin-indexing.cpp

```diff
@@ -824,55 +829,192 @@ namespace
 		using json = nlohmann::json;
 
 		try {
-			// skip the first line: @external
-			std::string rule_text{_rule_text};
-			if (_rule_text.find("@external") != std::string::npos) {
-				rule_text = _rule_text.substr(10);
+			log_re::debug("_rule_text => [{}]", _rule_text);
+
+			std::string_view rule_text = _rule_text;
+
+			// irule <text>
+			if (rule_text.find("@external rule {") != std::string_view::npos) {
+				const auto start = rule_text.find_first_of('{') + 1;
+				const auto end = rule_text.rfind(" }");
+
+				if (end == std::string_view::npos) {
+					auto msg = fmt::format("Received malformed rule text. "
+					                       "Expected closing curly brace following rule text [{}].",
+					                       rule_text);
+					log_re::error(msg);
+					return ERROR(SYS_INVALID_INPUT_PARAM, std::move(msg));
+				}
+
+				rule_text = rule_text.substr(start, end - start);
+			}
+			// irule -F <script>
+			else if (const auto external_pos = rule_text.find("@external\n"); external_pos != std::string_view::npos) {
+				// If there are opening and closing curly braces following the "@external\n" prefix, then we
+				// can assume that the rule text most likely represents a JSON string.
+				if (const auto start = rule_text.find_first_of('{'); start != std::string_view::npos) {
+					const auto end = rule_text.rfind(" }");
+
+					if (end == std::string_view::npos) {
+						auto msg = fmt::format("Received malformed rule text. "
+						                       "Expected closing curly brace following rule text [{}].",
+						                       rule_text);
+						log_re::error(msg);
+						return ERROR(SYS_INVALID_INPUT_PARAM, std::move(msg));
+					}
+
+					rule_text = rule_text.substr(start, end - start);
+				}
+				// Otherwise, the rule text must represent something else. In this case, simply strip the
+				// "@external\n" prefix from the rule text and let the JSON parser throw an exception if the
+				// rule text cannot be parsed. This allows the REP to fail without causing the agent to crash.
+				else {
+					rule_text = rule_text.substr(external_pos + 10);
+				}
 			}
```

Comment:
I've used this code in two plugins now. It probably needs to be provided by the irods-dev package so we avoid copying it everywhere.

Reply:
please make an issue - that seems good.

Reply:
I'm starting to think maybe this should be a documentation exercise. That code makes a few assumptions about the input which isn't code for general purpose use.

Will think on it a little more.